[docs] Deploy a python_function model on Microsoft Azure ML - guidance (#5491)
mlflow/mlflow: docs/source/models.rst

Deploy a ``python_function`` model on Microsoft Azure ML
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The MLflow plugin `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ can deploy models to Azure ML, either to Azure Kubernetes Service (AKS) or Azure Container Instances (ACI) for real-time serving. 

> The resulting container image contains a web server that accepts the following data formats as input:

Reviewer: Should we refer to the resulting deployment, rather than the resulting "container image containing a web server"?  The container image terminology was more relevant for the deprecated `mlflow.azureml.build_image` tool.

Author: Super agreed. Updated on next commit.

* JSON-serialized pandas DataFrames in the ``split`` orientation. For example, ``data = pandas_df.to_json(orient='split')``. This format is specified using a ``Content-Type`` request header value of ``application/json``.

> Deployments can be generated using both the Python API or MLflow CLI. On both cases, a ``JSON`` configuration file has to be specified with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI). You can check the full specification at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. The deployment configuration file looks as follows:

Reviewer: ```suggestion
Deployments can be generated using both the Python API or MLflow CLI. In both cases, a ``JSON`` configuration file has to be specified with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI). You can view the full specification at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. The deployment configuration file looks as follows:
```

A couple questions here:

1. The user doesn't actually have to pass a file, right? They can pass a dictionary?
2. We say that some JSON "has to be specified", but then we discuss a case where defaults are used when the JSON is not specified. Can we clarify?
3. The following JSON code blocks are just examples containing a subset of the available configuration options, right? Can we clarify that?

Author: 1. Yes, the user has to pass a file right now :(. In-place dictionaries are not supported since this is a wrapper on the client.
2. I got you point. I wanted to stress out the idea that if you don't indicate a JSON deployment file, one will be generated for you with some defaults. So yes, it is optional in the client, but no, the service needs one. I agree I had to make that more clear. Addressed in next commit.
3. I think It is stated on "You can view the full specification at [Deployment configuration schema](https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema). But I can make it more clear.

.. rubric:: For an ACI deployment

.. code-block:: json

    {

>       computeType: aci,

Reviewer 2: CircleCI checks are failing for invalid json structure. Here's an example that you can use to refer to: https://github.com/mlflow/mlflow/blob/master/docs/source/rest-api.rst#L664

Author: Goit it. Thanks for the handy example! It seems keys need to be in double quotes!

Reviewer 2: I typically debug docs changes locally by:
```
cd docs
make clean  # clears out the cache in '_build'
make html SPHINXOPTS="-W --keep-going -n"
```
There will be an absolute wall of text generated and the first run you do will take a solid 10 minutes to build the docker containers, but you'll get a pretty good idea of verbose reporting on issues without having to push changes to remote. 

That being said, sphinx is notorious for being hard to track down just WHY it's lexer can't parse the format of something... A lot of formatting for me is done by trying to find examples and doing a lot of trial and error. Spacing & indentation is key for formats to be accepted, so do keep that in mind.

Author: Thanks @BenWilson2 for the tips!

 * If ``location`` is not indicated, it defaults to the location of the workspace.

.. rubric:: For an AKS deployment

.. code-block:: json

    {
      computeType: aks,
      computeTargetName: aks-mlflow
    }

Remarks:
  * In above exmaple, ``aks-mlflow`` is the name of an Azure Kubernetes Cluster registered/created in Azure Machine Learning.

The following examples show how to create a deployment in ACI. Please, ensure you have `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ installed before continuing.

.. rubric:: Example: Workflow using the Python API

.. code-block:: py

    import json
    from mlflow.deployments import get_deploy_client

    # Create the deployment configuration.
    # If no deployment configuration is provided, then the deployment happens on ACI.
    deploy_config = {
        "computeType": "aci"
    }

    # Write the deployment configuration into a file.
    deployment_config_path = "deployment_config.json"
    with open(deployment_config_path, "w") as outfile:
        outfile.write(json.dumps(deploy_config))

    # Set the tracking uri as the deployment client.

>     client = get_deploy_client(mlflow.get_tracking_uri())

Reviewer: Is this the recommended best practice? What is the form of the MLflow Tracking URI when used on AzureML?

In general, can we discuss the format of the AzureML deployment URI for use with `mlflow deployments`? Some of the CLI examples below refer to `<azureml-mlflow-tracking-uri>`, but they don't describe the structure of this URI and don't tell users what other AzureML deployment targets they can use (e.g. options other than the MLflow Tracking URI, if any).

Author: Good point, the URL can be grabbed from the Azure portal because it used to be a dynamic URL (meaning it used to be valid for 1 hour since authentication was embedded on it). Now, the URL has an specific form and interactive autentication is triggered. I think we can explain that here. On Azure ML compute, the environment variable `MLFLOW_TRACKING_URI` is automatically populated with this tracking URL and no interactive authentication will happen as user is already authenticated. So, I think I can replace `mlflow.get_tracking_uri()` by `os.environ["MLFLOW_TRACKING_URI"]` and let the user either populate that manually or if using Azure ML compute, that is done for them. Makes sense?

    # webservice's scoring URI. The following example posts a sample input from the wine dataset
    # used in the MLflow ElasticNet example:
    # https://github.com/mlflow/mlflow/tree/master/examples/sklearn_elasticnet_wine
    # webservice's scoring URI.

    scoring_uri=$(az ml service show --name <deployment-name> -v | jq -r ".scoringUri")

    # `sample_input` is a JSON-serialized pandas DataFrame with the `split` orientation
    sample_input='
    {
        "columns": [
            "alcohol",
            "chlorides",
            "citric acid",
            "density",
            "fixed acidity",
            "free sulfur dioxide",
            "pH",
            "residual sugar",
            "sulphates",
            "total sulfur dioxide",
            "volatile acidity"
        ],
        "data": [
            [8.8, 0.045, 0.36, 1.001, 7, 45, 3, 20.7, 0.45, 170, 0.27]
        ]
    }'
You can also test your deployments locally first using the option `run-local`:

.. code-block:: bash

    echo $sample_input | curl -s -X POST $scoring_uri\
    -H 'Cache-Control: no-cache'\
    -H 'Content-Type: application/json'\

>     -d @-

Reviewer: We seem to be removing some valuable example content. Can we keep it around?

Author: I removed cause the requests are exactly the same as any request accepted by MLFlow serve, so this is no specific to Azure ML. I'd rather point users to that documentation. Does it make sense or do you think we should include it here too?

Reviewer 2: I think there's something to be said for having the example in place here. Let's add that back in so as to not have users have to context switch to a different location while reading through how to do this.

Author: Sure. I added them back on [73a18e6](https://github.com/mlflow/mlflow/pull/5491/commits/73a18e6198992ff818df2d0564632662cd2f169d).

Deploy a ``python_function`` model on Microsoft Azure ML
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The MLflow plugin `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ can deploy models to Azure ML, either to Azure Kubernetes Service (AKS) or Azure Container Instances (ACI) for real-time serving. 

The resulting deployment accepts the following data formats as input:

* JSON-serialized pandas DataFrames in the ``split`` orientation. For example, ``data = pandas_df.to_json(orient='split')``. This format is specified using a ``Content-Type`` request header value of ``application/json``.

.. warning::
    The ``TensorSpec`` input format is not fully supported for deployments on Azure Machine Learning at the moment. Be aware that many ``autolog()`` implementations may use ``TensorSpec`` for model's signatures when logging models and hence those deployments will fail in Azure ML.

> Deployments can be generated using both the Python API or MLflow CLI. In both cases, a ``JSON`` configuration file can be indicated with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI) and a minimal configuration. The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. Also, you will also need the Azure ML MLFlow Tracking URI of your particular Azure ML Workspace where you want to deploy your model. You can obtain this URI in several ways:

Reviewer: ```suggestion
Deployments can be generated using both the Python API or MLflow CLI. In both cases, a ``JSON`` configuration file can be indicated with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI) and a minimal configuration. The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. Also, you will also need the Azure ML MLflow Tracking URI of your particular Azure ML Workspace where you want to deploy your model. You can obtain this URI in several ways:
```

The MLflow plugin `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ can deploy models to Azure ML, either to Azure Kubernetes Service (AKS) or Azure Container Instances (ACI) for real-time serving. 

> The resulting deployment accepts the following data formats as input:

Reviewer 2: "format" 
`The resulting deployment accepts the following data format as input:`

Author: I got your point, but since it is a bullet list I think plural is right. It also gives the idea that there are other formats that are not supported. Does it make sense?

Reviewer 2: Yep! Makes sense. SGTM.

* JSON-serialized pandas DataFrames in the ``split`` orientation. For example, ``data = pandas_df.to_json(orient='split')``. This format is specified using a ``Content-Type`` request header value of ``application/json``.

> Deployments can be generated using both the Python API or MLflow CLI. On both cases, a ``JSON`` configuration file can be indicated with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI) and a minimal configuration. The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. On both cases, you will also need the Azure ML MLFlow Tracking URI of your particular Azure ML Workspace where you want to deploy your model. You can obtain this URI in several ways:

Author: Addressed on 
[540d0ee](https://github.com/mlflow/mlflow/pull/5491/commits/540d0eee60d4a7268b63c219e4a767dfa5c6c7da)

> * In the `Azure ML portal <https://ml.azure.com>`_, in the upper right corner, clic on the name of the workspace and on the pop-up blade that will show-up, clic on ``View all properties in Azure Portal``. In the properties panel, you will see the value corresponding to ``MLflow tracking URI``.

Reviewer 2: clic -> click

Also, could you convert this to a list of steps?
```
  * Through the `Azure ML portal <https://ml.azure.com>`:
    * Navigate to your Azure ML portal page.
    * Click on the name of the workspace at the upper right corner of the page.
    * Click "View all properties in Azure Portal" on the pane popup.
    * Copy the ``MLflow tracking URI`` value from the properties panel.

* Programmatically through the use of the ``Azure ML SDK``, using the method ....
```

Author: Addressed on [e8f336f](https://github.com/mlflow/mlflow/pull/5491/commits/e8f336f565706d7c1a180afbfd18ebd476c6d2e9)
